Add firmware_update support to c8y mapper

rina23q opened the pull request:
## Proposed changes
Implement `firmware_update` support to c8y-mapper as described in #2453.

## Types of changes

<!--
What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_
-->

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
#2453 

## Checklist

<!--
_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to ask.
We're here to help! This is simply a reminder of what we are going to look for
before merging your code._
-->

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

albinsuresh commented on crates/extensions/c8y_mapper_ext/src/firmware_update.rs:
```diff
@@ -0,0 +1,588 @@
+        .await
+        .expect("Send failed");
+
+        let (topic, received_json) = mqtt
+            .recv()
+            .await
+            .map(|msg| {
+                (
+                    msg.topic,
+                    serde_json::from_str::<serde_json::Value>(msg.payload.as_str().expect("UTF8"))
+                        .expect("JSON"),
+                )
+            })
+            .expect("No message received");
+
+        let (entity, channel) = MqttSchema::default().entity_channel_of(&topic).unwrap();
+        assert_eq!(entity, "device/main//");
+
+        match channel {
+            Channel::Command {
+                operation: OperationType::FirmwareUpdate,
+                ..
+            } => {
+                let expected_json = json!({
+                    "status": "init",
+                    "name": "myFirmware",
+                    "version": "1.0",
+                    "remoteUrl": "http://www.my.url"
+                });
+                assert_json_diff::assert_json_include!(actual: received_json, expected: expected_json);
+            }
+            _ => panic!("Unexpected response on channel: {channel:?}, topic: {topic:?}"),
+        }
+    }
```
So much code to just match a topic-prefix and the payload, just because the topic is dynamic with the cmd_id, highlights to the need of a helper function that matches topic prefix instead of the whole topic. That helper would be useful for all such command converter tests for other ops as well.

rina23q replied:
It was kinda my mistake. Since `firmware_update` doesn't produce `tedgeUrl`, so the JSON payload doesn't need to get `cmd_id` to assert. Here I replaced simply with `assert_received_includes_json`.

rina23q commented on crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs:
```diff
@@ -386,6 +386,10 @@ define_tedge_config! {
             config_update: bool,
+
+            /// Enable firmware_update feature
+            #[tedge_config(example = "true", default(value = false))]
```
Exactly, until operation workflow comes

rina23q commented on crates/extensions/c8y_mapper_ext/src/firmware_update.rs:
```diff
@@ -0,0 +1,588 @@
+        &self,
+        smartrest: &str,
+    ) -> Result<Vec<Message>, CumulocityMapperError> {
+        if !self.config.capabilities.firmware_update {
+            warn!(
+                "Received a c8y_Firmware operation, however, firmware_update feature is disabled"
+            );
+            return Ok(vec![]);
+        }
```
We must check capabilities in `process_smartrest`, otherwise, user cannot use custom operation for `c8y_Firmware` and other operations which are supported by thin-edge.
 
This is the reason why [firmware_operation.robot](https://github.com/thin-edge/thin-edge.io/blob/main/tests/RobotFramework/tests/cumulocity/firmware/firmware_operation.robot) failed in the very first try of this PR. The test uses custom operation for `c8y_Firmware` (`515`). If we don't have the capability check in `process_smartrest` level, the code would enter to this function, instead of calling the custom operation function.

albinsuresh replied:
Aah okay. Then these warnings also will have to be moved into that `process_smartrest` function, as they are completely useless here.

rina23q replied:
Removed by bfd52a307fbb6768db7308431544b040b30e2c9d

reubenmiller commented on crates/core/tedge_api/src/messages.rs:
```diff
@@ -616,6 +616,47 @@ impl ConfigUpdateCmdPayload {
+    pub name: Option<String>,
+    pub version: Option<String>,
+    #[serde(rename = "url")]
+    pub remote_url: Option<String>,
```
Generally if the url of the firmware binary is known then it should be included. If you are talking about publishing the current firmware info (name, version) on startup, then it is possible that the url is not known (as it might be baked into the OS image and was not provided via a url).

albinsuresh replied:
Since the `url` is useless in the initial bootstrap message, that reports the factory firmware info on the device, the only other case where it'd be useful is after a firmware update is performed. In that case, this `url` must be the one the device received in the cloud request (SR 515)? I don't see why that's needed either, because the `successful` status message already includes that `remoteUrl` field, which the mapper is using to send the `115` message. So, including the same `url` here seems redundant.

reubenmiller replied:
I don't have the context anymore of what message we're talking about.

reubenmiller replied:
But having the `url` property can't harm the message, and it is more descriptive about where the firmware came from (if the info is available).

albinsuresh replied:
> I don't have the context anymore of what message we're talking about. But more information can be useful..but without concrete examples of what messages we're explicitly talking about then I can't make an informed decision.

Okay, so here are the two cases where this "tedge firmware metadata" message would be sent:

1. When a device is getting bootstrapped: it needs to publish its factory firmware details. In this case, at best the device would know its `version` and `name`. So, the device is less likely to use the `url` field in this path.
2. After the device firmware is updated using a cloud request (with a remote URL): the `firmware_update` request would contain that `remoteUrl` and once this operation is successful, the device would send the `successful` command status message with the same `remoteUrl`. The mapper on receipt of this `successful` status will mark the forward that to the cloud (SR 503 message), followed by the SmartREST set firmware message (SR 115) where the same `remoteUrl` in the `successful` status response is used. Since the firmware version in the cloud is already updated this way, the device need not send the **tedge firmware metadata message** again. Now, on the next restart of the agent, this tedge firmware metadata message would be sent again, but here also, it won't be able to send that `url` information, unless it persists that URL info after every firmware update, which is less likely/unnecessary IMO.

This is the reason why I feel that the `url` field in the "tedge firmware metadata message" is completely useless and it is relevant only in the **firmware_update command** messages.

> But having the url property can't harm the message, and it is more descriptive about where the firmware came from (if the info is available).

Yeah, having an extra field doesn't harm, but when we document this API, we'll have to give some guidance to the user on when/how to use it, right? If this field is never going to be used, as neither cases highlighted above had a need for it (without additional persistence requirements), why even have it as an optional field?

reubenmiller replied:
I think we should still keep it as it "could" be useful to people if they want to implement it (e.g. workflow overriding etc.).

albinsuresh replied:
I'd have taken the opposite approach of omitting it for now and adding it when we encounter the need for it, as it's always easy to add things to a public API than removing things. Even for the workflow overriding, I only see the `remoteUrl` field in the firmware update `cmd` message being used and not this `url` field in the capability message.

But, since you're fairly confident that it would be useful in some use-cases, and since it's an optional field anyway, let's keep it for now. Once firmware update is fully implemented, and if this field is still not used, we can still remove it before 1.0.

reubenmiller commented on crates/core/tedge_api/src/messages.rs:
```diff
@@ -616,6 +616,47 @@ impl ConfigUpdateCmdPayload {
+#[serde(rename_all = "camelCase")]
+pub struct FirmwareMetadata {
+    pub name: Option<String>,
+    pub version: Option<String>,
```
Unless if the user does not know the name/version, or they would rather just use whatever name/version is actually contained inside the firmware image linked to the give url. Otherwise you would have to be sure that the name/version actually matched the whatever binary is behind the url.

I don't mind if name/version is optional (even though Cumulocity will not allow you to do this from the UI). But the local tedge interface should not be so strict as for us it does not really matter.

rina23q commented on crates/extensions/c8y_mapper_ext/src/lib.rs:
```diff
@@ -28,6 +30,7 @@ impl Default for Capabilities {
             log_upload: true,
             config_snapshot: true,
             config_update: true,
+            firmware_update: true,
```
Done 3b8290fa1154f6aeeb237b3ac5cfc47cbe9252ed

rina23q commented on crates/extensions/c8y_mapper_ext/src/firmware_update.rs:
```diff
@@ -0,0 +1,588 @@
+                self.ops_dir.clone().join(child_dir_name)
+            }
+            EntityType::Service => {
+                // No support for service firmware_update
```
Done 3b8290fa1154f6aeeb237b3ac5cfc47cbe9252ed

rina23q commented on crates/extensions/c8y_mapper_ext/src/firmware_update.rs:
```diff
@@ -0,0 +1,588 @@
+            "115,{name},{version},{url}",
+            name = metadata.name.unwrap_or("none".into()),
+            version = metadata.version.unwrap_or("none".into()),
+            url = metadata.remote_url.unwrap_or("none".into()),
```
Done 3b8290fa1154f6aeeb237b3ac5cfc47cbe9252ed

albinsuresh reviewed (commented):
The code changes LGTM. Only two outstanding comments to be discussed/resolved:

1. The need for `url` field in the firmware metadata message
2. Moving the cpability enabled check and warning into the `process_smartrest` function.

rina23q commented:
> Moving the cpability enabled check and warning into the process_smartrest function.
I removed the unreachable code by bfd52a3. However, removed the warnings since it was difficult to move warnings as there are possibilities that custom operation may catch the disabled SmartREST ID.

albinsuresh reviewed (approved):
LGTM
